src/main/java/io/stargate/sgv2/jsonapi/service/operation/model/ReadOperation.java

```diff
@@ -77,6 +90,146 @@ default Uni<FindResponse> findDocument(
+    final JsonNodeFactory nodeFactory = objectMapper.getNodeFactory();
+    return Multi.createBy()
+        .repeating()
+        .uni(
+            () -> new AtomicReference<String>(null),
+            stateRef -> {
+              return queryExecutor.executeRead(
+                  query, Optional.ofNullable(stateRef.get()), pageSize);
+            })
+        // Read document while pagingState exists, limit for read is set at updateLimit +1
+        .whilst(resultSet -> extractPagingStateFromResultSet(resultSet) != null)
+        .onItem()
+        .transformToUniAndMerge(
+            resultSet -> {
+              Iterator<QueryOuterClass.Row> rowIterator =
+                  resultSet.getRowsList().stream().iterator();
+              int remaining = resultSet.getRowsCount();
+              documentCounter.addAndGet(remaining);
+              while (--remaining >= 0 && rowIterator.hasNext()) {
+                ReadDocument document = null;
+                QueryOuterClass.Row row = rowIterator.next();
+                List<JsonNode> sortValues = new ArrayList<>(numberOfOrderByColumn);
+                for (int sortColumnCount = 0;
+                    sortColumnCount < numberOfOrderByColumn;
+                    sortColumnCount++) {
+                  int columnCounter =
+                      SORTED_DATA_COLUMNS + ((sortColumnCount) * SORT_INDEX_COLUMNS_SIZE);
+                  QueryOuterClass.Value value = row.getValues(columnCounter);
+                  if (!value.hasNull()) {
+                    sortValues.add(nodeFactory.textNode(Values.string(value)));
+                    continue;
+                  }
+                  columnCounter++;
+                  value = row.getValues(columnCounter);
+                  if (!value.hasNull()) {
+                    sortValues.add(nodeFactory.numberNode(Values.decimal(value)));
+                    continue;
+                  }
+                  columnCounter++;
+                  value = row.getValues(columnCounter);
+                  if (!value.hasNull()) {
+                    sortValues.add(nodeFactory.booleanNode(Values.bool(value)));
+                    continue;
+                  }
+                  columnCounter++;
+                  value = row.getValues(columnCounter);
+                  if (!value.hasNull()) {
+                    sortValues.add(nodeFactory.nullNode());
+                    continue;
+                  }
+
+                  sortValues.add(nodeFactory.missingNode());
+                }
+                // Create ReadDocument with document id, grpc value for doc json and list of sort
+                // values
+                document =
+                    ReadDocument.from(
+                        getDocumentId(row.getValues(0)), // key
+                        row.getValues(1), // Deserialized value of doc_json
+                        sortValues);
+                sortedData.add(document);
```

Comment:
non thread safe access

Reply:
Talked about this yesterday use `collect().in()` approach..

Reply:
Yes. Implemented in #308